app/uk/gov/hmrc/trusts/controllers/SubmissionDraftController.scala

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright 2020 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.trusts.controllers
+
+import javax.inject.Inject
+import play.api.libs.json._
+import play.api.mvc.{Action, AnyContent}
+import uk.gov.hmrc.trusts.controllers.actions.IdentifierAction
+import uk.gov.hmrc.trusts.models.requests.IdentifierRequest
+import uk.gov.hmrc.trusts.models.{RegistrationSubmissionDraft, RegistrationSubmissionDraftData, RegistrationSubmissionDraftPiece, RegistrationSubmissionDraftSetData, RegistrationSubmissionDraftStatus}
+import uk.gov.hmrc.trusts.repositories.RegistrationSubmissionRepository
+import uk.gov.hmrc.trusts.services.{AuditService, LocalDateTimeService}
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+class SubmissionDraftController @Inject()(submissionRepository: RegistrationSubmissionRepository,
+                                          identify: IdentifierAction,
+                                          auditService: AuditService,
+                                          localDateTimeService: LocalDateTimeService
+                                       ) extends TrustsBaseController {
+
+  def setSection(draftId: String, sectionKey: String): Action[JsValue] = identify.async(parse.json) {
+    implicit request => {
+      request.body.validate[RegistrationSubmissionDraftData] match {
+        case JsSuccess(draftData, _) =>
+          submissionRepository.getDraft(draftId, request.identifier).flatMap(
+            result => {
+              val draft: RegistrationSubmissionDraft = result match {
+                case Some(draft) => draft
+                case None => RegistrationSubmissionDraft(draftId, request.identifier, localDateTimeService.now, Json.obj(), None, Some(true))
+              }
```

Comment:
There's a lot going on in this controller method, is there a way we can refactor this into single responsible functions?

Reply:
It used to be that way, but there was a lot of chatter between the front and back end. If we don't mind having four API calls to the backend per page submit to send all the different pieces, then I can make it be that way. This felt simpler in terms of making sure the data was consistent (e.g. status and reg pieces actually match the user answers).

Reply:
By doing it all at once, it keeps an update atomic. Splitting the different pieces out risks the data becoming inconsistent.